File: rust-tool/src/configuration.rs
///////////////////////////////////////////////////////////////////////////////
// NAME:            configuration.rs
//
// AUTHOR:          Ethan D. Twardy <ethan.twardy@gmail.com>
//
// DESCRIPTION:     Configuration
//
// CREATED:         01/08/2022
//
// LAST EDITED:     01/08/2022
//
// Copyright 2022, Ethan D. Twardy
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
////

use std::collections::HashMap;
use std::default::Default;
use std::error::Error;
use std::fs::File;
use std::io::{BufReader, BufWriter};

use serde::{Serialize, Deserialize};

#[derive(Serialize, Deserialize)]
pub enum VolumeConfig {
    External {
        url: String,
        revision: String,
    },
}

#[derive(Serialize, Deserialize)]
pub struct Configuration {
    pub version: u8,
    pub volumes: HashMap<String, VolumeConfig>,
}

impl Default for Configuration {
    fn default() -> Configuration {
        Configuration { version: 1, volumes: HashMap::new() }
    }
}

pub fn load(file: &str) -> Result<Configuration, Box<dyn Error>> {
    let reader = BufReader::new(File::open(file)?);
    Ok(serde_yaml::from_reader::<BufReader<File>, Configuration>(reader)?)
}

pub fn store(file: &str, configuration: &Configuration) ->
    Result<(), Box<dyn Error>>
{
    let writer = BufWriter::new(File::create(file)?);
    Ok(serde_yaml::to_writer::<BufWriter<File>, Configuration>(
        writer, configuration)?)
}

///////////////////////////////////////////////////////////////////////////////

File: volumetric/source/configuration.c
///////////////////////////////////////////////////////////////////////////////
// NAME:            configuration.c
//
// AUTHOR:          Ethan D. Twardy <ethan.twardy@gmail.com>
//
// DESCRIPTION:     Logic for loading configuration file.
//
// CREATED:         01/17/2022
//
// LAST EDITED:     01/17/2022
//
// Copyright 2022, Ethan D. Twardy
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
////

#include <stdio.h>
#include <string.h>

#include <gobiserde/yaml.h>

#include <configuration.h>

const char* CONFIGURATION_CURRENT_VERSION = "1.0";

static int visit_mapping(yaml_deserializer* deser, void* user_data,
    const char* key)
{
    VolumetricConfiguration* config = (VolumetricConfiguration*)user_data;
    int result = 0;
    if (!strcmp("version", key)) {
        // Version check. The version affects whether we can fully deserialize
        // the configuration, so we do need to check it here.
        result = gobiserde_yaml_deserialize_string(deser,
            &config->version);
        if (0 >= result) {
            return result;
        } else if (strcmp(CONFIGURATION_CURRENT_VERSION, config->version)) {
            return -PARSE_VERSION_MISMATCH;
        }
    }

    else if (!strcmp("volume-directory", key)) {
        result = gobiserde_yaml_deserialize_string(deser,
            &config->volume_directory);
    }

    return result;
}

ParseResult volumetric_configuration_deserialize_yaml(yaml_deserializer* deser,
    VolumetricConfiguration* config)
{
    return gobiserde_yaml_deserialize_map(deser, visit_mapping, config);
}

///////////////////////////////////////////////////////////////////////////////
